File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/client/entity/model/ModelPart.java
package pize.tests.voxelgame.client.entity.model;

import pize.graphics.util.Shader;
import pize.graphics.vertex.Mesh;
import pize.math.util.EulerAngles;
import pize.math.vecmath.vector.Vec3f;
import pize.tests.voxelgame.client.control.camera.GameCamera;

public class ModelPart{
    
    private ModelPart parent;
    private final Mesh mesh;
    private final Pose pose, initialPose;
    private boolean show;
    
    public ModelPart(Mesh mesh){
        this.mesh = mesh;
        pose = new Pose();
        initialPose = new Pose();
        show = true;
    }
    
    
    public void render(GameCamera camera, Shader shader, String modelMatrixUniform){
        initialPose.updateMatrices(camera, initialPose);
        if(parent == null)
            pose.updateMatrices(camera, initialPose);
        else
            pose.updateMatrices(camera, initialPose, parent.getPose());
        
        if(shader == null || !show)
            return;
        shader.setUniform(modelMatrixUniform, pose.getModelMatrix());
        mesh.render();
    }
    
    
    public void setParent(ModelPart parent){
        this.parent = parent;
    }
    
    public Mesh getMesh(){
        return mesh;
    }
    
    
    public Pose getPose(){
        return pose;
    }
    
    public Pose getInitialPose(){
        return initialPose;
    }
    
    
    public void setInitialPose(float x, float y, float z, float yaw, float pitch, float roll){
        initialPose.getPosition().set(x, y, z);
        initialPose.getRotation().set(yaw, pitch, roll);
    }
    
    public void setInitialPose(Vec3f position, EulerAngles rotation){
        setInitialPose(position.x, position.y, position.z, rotation.yaw, rotation.pitch, rotation.roll);
    }
    
    public void setInitialPose(double x, double y, double z){
        initialPose.getPosition().set(x, y, z);
    }
    
    public void setInitialPose(Vec3f position){
        setInitialPose(position.x, position.y, position.z);
    }
    
    public Vec3f getPosition(){
        return pose.getPosition();
    }
    
    public EulerAngles getRotation(){
        return pose.getRotation();
    }
    
    public boolean isShow(){
        return show;
    }
    
    public void setShow(boolean show){
        this.show = show;
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/client/entity/model/ModelPart.kt
package pize.tests.voxelgame.client.entity.model

import pize.graphics.util.Shader
import pize.graphics.vertex.Mesh
import pize.math.util.EulerAngles
import pize.math.vecmath.vector.Vec3f
import pize.tests.voxelgame.client.control.camera.GameCamera

class ModelPart(val mesh: Mesh?) {
    private var parent: ModelPart? = null
    val pose: Pose
    val initialPose: Pose
    var isShow: Boolean

    init {
        pose = Pose()
        initialPose = Pose()
        isShow = true
    }

    fun render(camera: GameCamera?, shader: Shader?, modelMatrixUniform: String?) {
        initialPose.updateMatrices(camera, initialPose)
        if (parent == null) pose.updateMatrices(camera, initialPose) else pose.updateMatrices(
            camera,
            initialPose,
            parent!!.pose
        )
        if (shader == null || !isShow) return
        shader.setUniform(modelMatrixUniform, pose.modelMatrix)
        mesh!!.render()
    }

    fun setParent(parent: ModelPart?) {
        this.parent = parent
    }

    fun setInitialPose(x: Float, y: Float, z: Float, yaw: Float, pitch: Float, roll: Float) {
        initialPose.position[x, y] = z
        initialPose.rotation[yaw, pitch] = roll
    }

    fun setInitialPose(position: Vec3f, rotation: EulerAngles) {
        setInitialPose(position.x, position.y, position.z, rotation.yaw, rotation.pitch, rotation.roll)
    }

    fun setInitialPose(x: Double, y: Double, z: Double) {
        initialPose.position[x, y] = z
    }

    fun setInitialPose(position: Vec3f) {
        setInitialPose(position.x.toDouble(), position.y.toDouble(), position.z.toDouble())
    }

    val position: Vec3f?
        get() = pose.position
    val rotation: EulerAngles?
        get() = pose.rotation
}
